flecsi/data.hh

```diff
@@ -31,18 +30,20 @@
 #include <flecsi/data/accessor.hh>
 #include <flecsi/data/coloring.hh>
 #include <flecsi/data/privilege.hh>
+#include "flecsi/data/topology_accessor.hh"
+#include "flecsi/util/export_definitions.hh"
 
 namespace flecsi {
 
 /*
   Default global topology instance.
  */
-inline topo::global::slot global_topology;
+FLECSI_EXPORT extern topo::global::slot global_topology;
```

Comment:
I know Windows doesn't implement C++, but some of the details elude me.  I've seen header files decorated with lots of `__declspec(dllexport)` or so, usually with a preprocessor switch such that they can be reinterpreted as `__declspec(dllimport)` in clients; are you doing that here?  Moreover, we have lots of inline variables, mostly to declare `field<…>::definition` objects, so we surely, eh, need more of this.

Reply:
> I know Windows doesn't implement C++, but some of the details elude me. I've seen header files decorated with lots of `__declspec(dllexport)` or so, usually with a preprocessor switch such that they can be reinterpreted as `__declspec(dllimport)` in clients; are you doing that here? 

Yes, that's what's happening here.

Windows visibility rules for symbols defined in shared libraries are different from those on Linux (they are similar to what MacOS is doing, btw): 

- symbols need to be explicitly marked as exported in order for them to be usable from other binary modules (the `FLECSI_EXPORT` macro does the right thing on all platforms - it expands to `__declspec(dllimport)`/`__declspec(dllexport)` on Windows and to `__attribute__((visibility("default")))` on Linux (see: https://github.com/flecsi/flecsi/blob/devel/flecsi/util/export_definitions.hh#L16-L34)
- as a consequence, on Windows every binary module that sees an `inline` (in fact, any global) variable will instantiate it's own copy of this variable. This causes issues as code that assumes to operate on the same instance of a global variable ends up being broken. 

> Moreover, we have lots of inline variables, mostly to declare field<…>::definition objects, so we surely, eh, need more of this.

Possibly, yes. We might need similar changes for (inline) variables that have a state.

Reply:
My understanding is that macOS does a similar but different thing involving multiple namespaces in which each dynamic object can define symbols.  In any case, it's highly unfortunate to not have the language behavior even as an _option_ (even if, in its turn, the language does not address dynamic linking at all as of yet).

Reply:
@opensdh From what I see now, it will not be possible to have the code base properly working as is on Windows after all. As you mentioned you're using a number of inline variables and a couple of Meyer's singletons. As long as those are being used in non-template types, the `FLECSI_EXPORT` can be applied. However, for template types this would require explicit instantiation, which is not what we want. Looks like that I will have to move my development environment to a Linux system. Please stop reviewing this for the time being, I will roll back the changes specific to MSVC/Windows before continuing.

Reply:
As little as I'm interested personally in supporting Windows, maybe it's not quite that bad: several templated field definitions like `buffers_category<P>::field`, `narray<P>::meta_field`, and `ntree<P>::e_coordinates` are inline, but my suspicion is that the FleCSI dynamic object contains few if any instances of those templated variables (since it defines only a few types that can serve as `P`), so perhaps after a _few_ explicit instantiations only the application needs to worry about making its symbols unique.  `identity_traits<R>::value` is inline, but its, er, _identity_ probably doesn't matter.  The same goes for `fold::wrap<R,T>::id` (or `…::op` in the MPI backend), with only a little waste if we register effectively the same reduction function more than once.  The only other templated inline variables that I see are the static `std::string` objects in the classes derived from `annotation::execute_task`, which presumably work well enough after you changed them to be Meyer's singletons.

Reply:
I think the main problem are not the types you mentioned. The use case that made me give up is this function: https://github.com/flecsi/flecsi/blob/devel/flecsi/topo/core.hh#L126-L129. It gets instantiated with various, non-obvious types and I don't think we can control the module that is instantiating it.

Reply:
Aha.  I looked at the inline variables because you changed some, but the inline functions with locals are indeed just the same.  Indeed, that identifier is used variously to register fields and to retrieve the field registrations, and it would be hard to partition that between FleCSI and its clients.  Should we consider linking statically (on Windows)?

In general, my apologies for my inability to write "Windows C++" (and/or "macOS C++", although I (still) think that case is configurable).

Reply:
Yes, static linking might be an option, moreover as the core FleCSI shared library is very small.

Reply:
@opensdh statically linking did indeed work, thanks again for suggesting.